File: PaperLib/Repositories/Exporters/Exporter.swift
//
//  Exporter.swift
//  PaperLib
//
//  Created by GeoffreyChen on 16/02/2022.
//

import Foundation
import Alamofire
import Combine
import SwiftUI

enum ExportFormat {
    case plain
    case bibtex
}

class Exporter {
    func export(entities: [PaperEntity], format: ExportFormat) {
        switch format {
        case .plain:
            return self._exportPlain(entities: entities)
        case .bibtex:
            return self._exportBibtex(entities: entities)
        }
    }

    func _replacePublication(_ publication: String) -> String {
        do {
            if let exportReplacement = UserDefaults.standard.data(forKey: "exportReplacement"), UserDefaults.standard.bool(forKey: "enableExportReplacement") {
                let exportReplacementMap = try JSONDecoder().decode([String: String].self, from: exportReplacement)
                if let replace = exportReplacementMap[publication] {
                    return replace
                } else {
                    return publication
                }
            } else {
                return publication
            }
        } catch let err {
            print(err)
            return publication
        }
    }
}

extension Exporter {
    func _exportPlain(entities: [PaperEntity]) {
        var allPlain = ""

        entities.forEach { entity in
            let text = "\(entity.authors). \"\(entity.title)\" In \(_replacePublication(entity.publication)), \(entity.pubTime). \n\n"
            allPlain += text
        }
        let pasteBoard = NSPasteboard.general
        pasteBoard.clearContents()
        pasteBoard.writeObjects([allPlain as NSString])
    }
}

extension Exporter {
    func _exportBibtex(entities: [PaperEntity]) {
        var allTexBib = ""

        entities.forEach { entity in
            var citeKey = ""
            citeKey += entity.authors.split(separator: " ")[0] + "_"
            citeKey += entity.pubTime + "_"
            citeKey += formatString(String(entity.title[..<String.Index(utf16Offset: 3, in: entity.title)]), removeNewline: true, removeWhite: true, removeSymbol: true)!
            var texbib = ""
            if entity.pubType == 1 {
                texbib = """
                @inproceedings{\(citeKey),
                    year = \(entity.pubTime),
                    title = {\(entity.title)},
                    author = {\(entity.authors.replacingOccurrences(of: ", ", with: " and "))},
                    booktitle = {\(_replacePublication(entity.publication))},
                }
                """
            } else {
                texbib = """
                @article{\(citeKey),
                    year = \(entity.pubTime),
                    title = {\(entity.title)},
                    author = {\(entity.authors.replacingOccurrences(of: ", ", with: " and "))},
                    journal = {\(_replacePublication(entity.publication))},
                }
                """
            }
            allTexBib += texbib + "\n\n"
        }
        let pasteBoard = NSPasteboard.general
        pasteBoard.clearContents()
        pasteBoard.writeObjects([allTexBib as NSString])
    }
}

File: src/repositories/exporters/exporter.js
import {formatString} from '../../utils/misc';

export class Exporter {
  constructor(preference) {
    this.preference = preference;
  }

  export(entities, format) {
    if (format === 'bibtex') {
      return this._exportBibtex(entities);
    } else if (format === 'plain') {
      return this._exportPlainText(entities);
    }
  }

  _replacePublication(publication) {
    if (this.preference.get('enableExportReplacement')) {
      for (const kv of this.preference.get('exportReplacement')) {
        if (kv.from == publication) {
          return kv.to;
        }
      }
    }
    return publication;
  }

  _exportBibtex(entities) {
    let allTexBib = '';

    for (const entity of entities) {
      let citeKey = '';
      citeKey += entity.authors.split(' ')[0] + '_';
      citeKey += entity.pubTime + '_';
      citeKey += formatString({
        str: entity.title.slice(0, 3),
        removeNewline: true,
        removeWhite: true,
        removeSymbol: true,
      });
      let texbib = '';
      if (entity.pubType == 1) {
        texbib = `@inproceedings{${citeKey},
    year = ${entity.pubTime},
    title = {${entity.title}},
    author = {${entity.authors.replace(', ', ' and ')}},
    booktitle = {${this._replacePublication(entity.publication)}},
}`;
      } else {
        texbib = `@article{${citeKey},
    year = ${entity.pubTime},
    title = {${entity.title}},
    author = {${entity.authors.replace(', ', ' and ')}},
    journal = {${this._replacePublication(entity.publication)}},
}`;
      }
      allTexBib += texbib + '\n\n';
    }
    return allTexBib;
  }

  _exportPlainText(entities) {
    let allPlain = '';

    for (const entity of entities) {
      const text = `${entity.authors}. \"${entity.title}\" In ${entity.publication}, ${entity.pubTime}. \n\n`;
      allPlain += text;
    }
    return allPlain;
  }
}
